File: Java/Hard/WrongAnswer/sum-of-total-strength-of-wizards_2.java
class Solution {
    public int totalStrength(int[] strength) {
        
        int n = strength.length;
        int[] left = new int[n];
        int[] right = new int[n];
        Stack<Integer> stack = new Stack<>();
        for(int i=0; i<n; i++){
            while(!stack.isEmpty() && strength[stack.peek()] > strength[i]){
                right[stack.pop()] = i;
            }
            left[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(i);
        }
        while(!stack.isEmpty()){
            right[stack.pop()] = n;
        }
        long mod = 1000000007;
        long sum = 0;
        for(int i=0; i<n; i++){
            long leftCount = i - left[i];
            long rightCount = right[i] - i;
            sum += strength[i] * leftCount * rightCount;
            sum %= mod;
        }
        return (int)sum;
    }
}

File: Python3/Hard/WrongAnswer/sum-of-total-strength-of-wizards_2.py
class Solution:
    def totalStrength(self, strength: List[int]) -> int:
        
        # all contiguous groups of wizards
        # total strength = min * sum
        # return the sum of all total strengths
        # brute force: O(n^2)
        # stack: O(n)
        # 1. find the min in the array
        # 2. find the sum of the array
        # 3. multiply min * sum
        # 4. add to the result
        # 5. pop the min from the array
        # 6. repeat 1-5 until the array is empty
        # return the result
        res = 0
        while strength:
            m = min(strength)
            s = sum(strength)
            res += m * s
            strength.remove(m)
        return res